feat(core) New instruments API (#2430)
n1ru4l/envelop: website/src/pages/docs/plugins/lifecycle.mdx

- `schema` - the `GraphQLSchema`
- `replaceSchema` - replaces the schema. Calling this will trigger `onSchemaChange` for all other
  plugins (except for the one that initiated the change);

### `instruments`

An optional `instruments` instance can be present in the plugin.

> This `Instruments` instance allows to wrap an entire phase execution (including all plugin hooks),

Reviewer: ```suggestion
This `Instruments` instance allows wrapping an entire phase execution (including all plugin hooks);
```

> meaning running code just before, just after and around the execution of the phase.

Reviewer: ```suggestion
meaning, running code just before, just after and around the execution of the phase.
```

> Instruments doesn't have access to input/output of a phase, use hooks to have access to those data.

Reviewer: ```suggestion
Instruments don't have access to the input and outputs of a phase, use hooks to access that data.
```

If needed, we recommend to share data between instruments and hooks with a `WeakMap` and the given
`context` as the key.

> All instruments takes 2 parameters:

Reviewer: ```suggestion
All instruments take 2 parameters:
```

- `payload`: an object containing the graphql context.
- `wrapped`: The function representing the execution of the phase. It takes no parameters, and

>   returns `void` (or `Promise<void>` for asynchrone phases). **This function must always be

Reviewer: ```suggestion
  returns `void` (or `Promise<void>` for asynchronous phases). **This function must always be
```

  called**.

#### Instruments composition

> If multiple plugins have `instruments`, they are composed in the same order they are defined the

Reviewer: ```suggestion
If multiple plugins have `instruments`, they are composed in the same order as they are defined in the
```

> plugin array (the first is outtermost call, the last is inner most call).

Reviewer: ```suggestion
plugin array (from top to bottom).
```

> It is possible to customize this composition if it doesn't suite your need (ie. you need hooks and

Reviewer: ```suggestion
It is possible to customize this composition if it doesn't suit your needs (ie. you need hooks and
```

> instruments to have a different oreder of execution).

Reviewer: ```suggestion
instruments to have a different order of execution).
```

```ts
import { composeInstruments, envelop } from '@envelop/core'

const { instruments: instruments1, ...plugin1 } = usePlugin1()
const { instruments: instruments2, ...plugin2 } = usePlugin2()

const instruments = composeInstruments([instruments2, instruments1])

const getEnveloped = envelop({
  plugin: [{ insturments }, plugin1, plugin2]
})
```

#### `init`

Wraps the envelop (the call to `envelop` function) initialisation.

This includes all the plugins `onEnveloped` hooks, and the creation of the Envelop Orchestrator.

> This instruments must be synchrone, the wrapped function is always synchrone.

Reviewer: ```suggestion
This instrument must be synchronous, the wrapped function is always synchronous.
```

#### `parse`

> Wraps the parse phase. This includes all the plugins `onParse` hooks and the actual engine `parse`.

Reviewer: "engine" might be confusing.
```suggestion
Wraps the parse phase. This includes all the plugins `onParse` hooks and the actual `parse` function.
```

> This instrument must be synchrone, the wrapped function is always synchrone.

Reviewer: ```suggestion
This instrument must be synchronous, the wrapped function is always synchronous.
```

#### `validate`

> Wraps the validate phase. This includes all the plugins `onValidate` hooks and the actual engine
> `validate`.

Reviewer: ```suggestion
Wraps the validate phase. This includes all the plugins `onValidate` hooks and the actual `validate`
function.
```

> This instrument must be synchrone, the wrapped function is always synchrone.

Reviewer: ```suggestion
This instrument must be synchronous, the wrapped function is always synchronous.
```

#### `context`

Wraps the context building phase. This includes all the plugins `onContextBuilding` hooks.

> This instrument must be synchrone, the wrapped function is always synchrone.

Reviewer: ```suggestion
This instrument must be synchronous, the wrapped function is always synchronous.
```

#### `execute`

Wraps the execute phase. This includes all the plugins `onExecute` hooks.

> This instrument can be asynchrone, the wrapped function **can be** asynchrone. Be sure to `await` or

Reviewer: ```suggestion
This instrument can be asynchronous, the wrapped function **can be** asynchronous. Be sure to `await` or
```

use `.then` on the result of the `wrapped` function to run code after the `execute` phase.

Note that `wrapped` is not guaranted to return a promise.

#### `subscribe`

Wraps the subscribe phase. This includes all the plugins `onSubsribe` hooks. Note that it doesn't
wrap the entire lifetime of the subscription, but only it's intialisation.

> This instrument can be asynchrone, the wrapped function **can be** asynchrone. Be sure to `await` or

Reviewer: ```suggestion
This instrument can be asynchronous, the wrapped function **can be** asynchronous. Be sure to `await` or
```
